### ISSUE TITLE
machinectl enable doesn't work, nor first time machinectl start is run neither. However, second time it does

### ISSUE THREAD

**Issue by q2dg** [systemd/systemd#13130]
**systemd version the issue has been seen with**
> ...v241-8.git9ef65cb.fc30

**Used distribution**
 > …Fedora 30 (x86_64)

**Expected behaviour you didn't see**
 > …I want to have a container running since host boot. It seems easy executing _machinectl enable myCont_ but I notice it doesn't work on my machine. What is more, if I manually execute _machinectl start myCont_ the container doesn't run neither. But,if a rerun this last command again, then it starts flawessly. Looking at journald, the error I get is this:
Failed to set up loopback block device: Operation not permitted

I suspect it's something related to the order in (auto?)mounting some underlying devices but my skills in this topic are really bare.

Thanks

**Comment by poettering** [comment]
please provide logs of this!

**Comment by q2dg** [comment]
Well, there's no many information to log, unfortunately:
1.-When I boot host machine (when machinectl enable should work) there's no rellevant output in journalctl -xe
2.-When I execute machinectl start the first time, I get the error I've said. Here's a screenshot 

![Captura de pantalla de 2019-07-22 17-07-37](https://user-images.githubusercontent.com/7274874/61644250-a2f3a000-aca4-11e9-905d-f2f0d35bbf26.png)

3.-When I execute machinectl start the second time, I can see the normal and correct systemd's container boot messages in journalctl -xe . The rellevant difference is loop module kernel is correctly used, as it can be seen in next screenshot (it recognized "p1" and "p2" because my container have been made with "mkosi -b ...").

![Captura de pantalla de 2019-07-22 17-21-47](https://user-images.githubusercontent.com/7274874/61644687-6bd1be80-aca5-11e9-95db-93b0987f2a47.png)

So it seems a problem with the loading of the loop module kernel ... but I don't know why it happens

PD: I have SELinux disabled.

**Comment by poettering** [comment]
hmm, so my educated guess is this:

it's a race between the DeviceAllow= lines we ship in systemd-nspawn@.service. We have:

```
DeviceAllow=/dev/net/tun rwm
DeviceAllow=char-pts rw
DeviceAllow=/dev/loop-control rw
DeviceAllow=block-loop rw
DeviceAllow=block-blkext rw
DeviceAllow=/dev/mapper/control rw
DeviceAllow=block-device-mapper rw
```

these can only be resolved if the device nodes /dev/net/tun, /dev/loop-control or /dev/mapper/control exist and the block device subsystems "pts", "loop", "blkext" and "device-mapper" are loaded in the kernel, i.e. show up in /proc/devices... If these device nodes/subsystems aren't around the DeviceAllow= syntax has no effect unfortunately...

And in your case loop.ko is not compiled into the kernel, hence these cannot be resolved before nspawn runs, and then nspawn runs and triggers them to be loaded, but then cannot access them. On the second try the policy can be resolved, so the second invocation works fully.

Can you maybe do `systemctl edit systemd-nspawn@fedora.service`. That opens an editor, please there:
```
[Service]
ExecStartPre=-/usr/sbin/modprobe loop tun dm-mod
```

then try again. This should fix your issue. Please report back.

If this works I figure we could add a logic to the device allow logic to automatically issue the equivalent of `modprobe devname:x/y` for any device `/dev/y/y` listed in the access list. It would solve this automatically, and only be moderately ugly.

**Comment by poettering** [comment]
(the idea with modprobe devname:x/y doesn't work, since kmods that expose alias information that way already get their own device nodes created by tmpfiles.d snippets. this means they are resolvable anyway. which means what is missing here is that block-loop, block-device-mapper are resolvable, and I am not sure how we can do that without the explicit modprobe line suggested above)

**Comment by poettering** [comment]
I prepped a hack along these lines in #13144. Please have a look, and test if this works.

**Comment by q2dg** [comment]
Ok. I've added the line "ExecStartPre=-/sbin/modprobe -abq tun loop dm-mod" shown in PR to my /lib/systemd/system/systemd-nspawn@.service file but, however I've not touched my /lib/systemd/system/systemd-logind.service file (also modified in PR) because mine (from stock Fedora 30) has a very different content than the one used in PR  (in fact, it hasn't any DeviceAllow= line, for instance!... maybe it's because I'm using v241?). Anyway, I've reboot and...the error continues: nor _machinectl enable_ works nor _machinectl start_ works the first time. So I think this PR should be tested using last systemd version to be sure  (which I can't run). ...systemd-logind.service file was really different
Sorry :-(

**Comment by poettering** [comment]
can you turn on debug logging ("systemd.log_level=debug" on the kernel cmdline) and then paste the relevant output in the logs here. i.e. everything around the failing nspawn invocation?

**Comment by q2dg** [comment]
Of course!
Here it is! complete log from boot until the first execution of "machinectl start fedora" (time 15:37:20) and the second one (time 15:41:00). Line "ExecStartPre=-/sbin/modprobe ..." remains written.
Thanks a lot
[machinectl-log.txt](https://github.com/systemd/systemd/files/3427076/machinectl-log.txt)

**Comment by poettering** [comment]
that's without debug logging... please boot with "system.log_level=debug" on the kernel cmdline and then regenerate

**Comment by q2dg** [comment]
Oooh, sorry, I forgot to do the grub2-mkconfig after editing GRUB_CMDLINE in /etc/default/grub. Sorry!!! Here it is. First invocation to _machinectl start_ is at 23:22:09, second one is at 23:22:10.

[machinectl-log.txt](https://github.com/systemd/systemd/files/3428819/machinectl-log.txt)

Just for the record (I forgot to say before although I think it's irrellevant), my /etc/systemd/nspawn/fedora.nspawn file is like this:

_[Exec]
PrivateUsers=false
[Network]
Private=no_

**Comment by q2dg** [comment]
Well, I just want to remark that I have the same error simply running machinectl start image.raw in Fedora 31 (systemd v243.4-1.fc31): first time it raises error but second time it works ok

**Comment by fooker** [comment]
Got the same problem on ArchLinux using `systemd 244 (244-1-arch)`. Will enable debug log on the next reboot cycle and try to provide additional infos.

**Comment by poettering** [comment]
Pretty sure this is fixed now with 625077264ba01a108386eeea733ee244e6b7ff14

Closing.

**Comment by bnf** [comment]
> Pretty sure this is fixed now with 6250772

Just for reference:
I faced the exact same problem as reported by @q2dg on a Fedora 30 machine (systemd v241-14.git18dd3fb.fc30) and I can confirm that 6250772 fixes the issue.
Thanks.
